test/e2e/util.go

```diff
@@ -1316,3 +1317,69 @@ func matchIPv4StringFamily(ipStrings []string) (string, error) {
 func matchIPv6StringFamily(ipStrings []string) (string, error) {
 	return util.MatchIPStringFamily(true /*ipv6*/, ipStrings)
 }
+
+// This is a replacement for e2epod.DeletePodWithWait(), which does not handle pods that
+// may be automatically restarted (https://issues.k8s.io/126785)
+func deletePodWithWait(ctx context.Context, c clientset.Interface, pod *v1.Pod) error {
+	if pod == nil {
+		return nil
+	}
+	if pod.ResourceVersion == "" {
+		// We only recurse into deletePodWithWaitByName when ResourceVersion is
+		// *not* set, to avoid infinite loops.
+		return deletePodWithWaitByName(ctx, c, pod.Name, pod.Namespace)
+	}
+	oldResourceVersion := pod.ResourceVersion
+
+	framework.Logf("Deleting pod %q in namespace %q", pod.Name, pod.Namespace)
+	err := c.CoreV1().Pods(pod.Namespace).Delete(ctx, pod.Name, metav1.DeleteOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil // assume pod was already deleted
+		}
+		return fmt.Errorf("pod Delete API error: %w", err)
+	}
+	framework.Logf("Wait up to %v for pod %q to be fully deleted", e2epod.PodDeleteTimeout, pod.Name)
+	err = waitForPodNotFoundInNamespace(ctx, c, pod.Name, pod.Namespace, oldResourceVersion, e2epod.PodDeleteTimeout)
+	if err != nil {
+		return fmt.Errorf("pod %q was not deleted: %w", pod.Name, err)
+	}
+	return nil
+}
+
+// This is a replacement for e2epod.DeletePodWithWaitByName(), which does not handle pods
+// that may be automatically restarted (https://issues.k8s.io/126785)
+func deletePodWithWaitByName(ctx context.Context, c clientset.Interface, podName, podNamespace string) error {
+	pod, err := c.CoreV1().Pods(podNamespace).Get(ctx, podName, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil // assume pod was already deleted
+		}
+		return fmt.Errorf("pod Get API error: %w", err)
+	}
+	if pod.ResourceVersion == "" {
+		return fmt.Errorf("unexpected Pod with no ResourceVersion returned from API!")
+	}
+	// We only recurse into deletePodWithWait when ResourceVersion *is* set, to avoid
+	// infinite loops.
+	return deletePodWithWait(ctx, c, pod)
+}
+
+// This is an alternative version of e2epod.WaitForPodNotFoundInNamespace(), which takes
+// a resourceVersion as well.
+func waitForPodNotFoundInNamespace(ctx context.Context, c clientset.Interface, podName, ns, resourceVersion string, timeout time.Duration) error {
+        err := framework.Gomega().Eventually(ctx, framework.HandleRetry(func(ctx context.Context) (*v1.Pod, error) {
+                pod, err := c.CoreV1().Pods(ns).Get(ctx, podName, metav1.GetOptions{})
+                if apierrors.IsNotFound(err) {
+                        return nil, nil
+                }
+		if pod != nil && pod.ResourceVersion != resourceVersion {
```

Comment:
by this stage we have submitted a delete request for the pod - can the resource version for the pod change? If it can, then we may falsely think its deleted here.

Reply:
no, ResourceVersion is immutable; if ResourceVersion has changed, that means the pod was deleted and recreated. (This check is the difference from the original `e2epod.WaitForPodNotFoundInNamespace`.)

Reply:
I thought resource version implemented by k8 may increment with each change applied to the pod object. Youre right that if it was deleted and created it would get a different resource version but it can also change with any changes to the pod object.

Reply:
uh... wow, not sure what I was thinking...